File: 2065. Maximum Path Quality of a Graph/Solution.py
from typing import List


class Solution:
    def maximalPathQuality(self, values: List[int], edges: List[List[int]], maxTime: int) -> int:
        n = len(values)
        graph = [[] for _ in range(n)]
        for x, y, t in edges:
            graph[x].append((y, t))
            graph[y].append((x, t))

        res = 0
        visited = [0] * n

        def dfs(x: int, remainingTime: int, curValues: int) -> None:
            if visited[x] == 0:
                curValues += values[x]
            visited[x] += 1

            if x == 0:
                nonlocal res
                res = max(res, curValues)

            for y, t in graph[x]:
                if t <= remainingTime:
                    dfs(y, remainingTime - t, curValues)

            visited[x] -= 1
            if visited[x] == 0:
                curValues -= values[x]

        dfs(0, maxTime, 0)

        return res

File: 2065. Maximum Path Quality of a Graph/Solution.cpp
#include <functional>
#include <vector>
using namespace std;

class Solution {
public:
    int maximalPathQuality(vector<int>& values, vector<vector<int>>& edges, int maxTime)
    {
        int n = values.size();
        vector<vector<pair<int, int>>> graph(n);
        for (const vector<int>& e : edges) {
            graph[e[0]].emplace_back(e[1], e[2]);
            graph[e[1]].emplace_back(e[0], e[2]);
        }

        int res = 0;
        vector<int> visited(n, 0);
        function<void(int, int, int)> dfs = [&](int node, int remainingTime, int curValues) -> void {
            if (visited[node] == 0)
                curValues += values[node];
            visited[node]++;

            if (node == 0)
                res = max(res, curValues);

            for (const auto& [neighbour, cost] : graph[node])
                if (cost <= remainingTime)
                    dfs(neighbour, remainingTime - cost, curValues);

            visited[node]--;
            if (visited[node] == 0)
                curValues -= values[node];
        };

        dfs(0, maxTime, 0);
        return res;
    }
};
